applications/chain-extension-suite.md

```diff
@@ -0,0 +1,240 @@
+
+### `pallet_assets_extensions` and `pallet_nft_extensions`
+
+The Substrate framework's FRAME provides `pallet_assets` for fungible tokens and `pallet_nft` for non-fungible tokens (NFTs), aligning with the prevalent token standards in the blockchain domain. On the smart contract side, ink! adheres to PSP-22 for fungible tokens and PSP-34 for NFTs.
+
+Within the objective of this grant we develop `pallet_assets_extensions` and `pallet_nft_extensions`. These extensions aim to expose chain extension functionality for `pallet_assets` and `pallet_nft`, thereby creating a bridge for enhanced interaction between the Substrate runtime and the smart contract environment of ink!.
+
+The envisioned extensions will include ready-to-go contracts for the ink! ecosystems, facilitating a seamless integration and interaction between the substrate runtime and smart contracts, adhering to the respective token standards. This initiative seeks to automate and standardize the chain extension process for fungible and non-fungible tokens within the Substrate ecosystem, thereby fostering an environment of ease and efficiency for developers working across these platforms.
+
+It specifically includes:
+- **PSP22 & PSP34 chain extensions** for `pallet_asset` and `pallet_nft`, ready to add to the `pallet_contract` runtime config type `ChainExtension`.
+- **PSP22 & PSP34 traits** written in ink.
+- **An asset and an NFT chain-extension** for ink! satisfying the full feature set of the respective standard alongside with actual implementation of the smart contracts in ink!.
+
+
+```rust
+
+/// Sample of the final PSP22 code …. 
+impl PSP22 for AssetContract {
+
+  /// Transfer tokens from the caller to the `to` account.
+  #[ink(message)]
+  fn transfer(
+    &mut self, 
+    to: AccountId, 
+    value: Balance,
+    _data: Vec<u8>
+  ) -> Result<(), PSP22Error> {
+     let result = self
+       .env()
+       .extension()
+       .transfer(self.asset_id, to, value)
+       .map_err(PSP22Error::from);
+    
+     if result.is_ok() {
+       self.env().emit_event(Transfer {
+         from: Some(self.env().caller()),
+         to: Some(to),
+        value,
+       });
+     }
+     result
+  }
+}
+```
+
+### ink! v5 Upgrade
+
+The forthcoming release of ink! v5 is anticipated to bring about enhanced features and optimizations. Being at the forefront of these developments, the Chain Extension Suite is designed to be compatible with ink! v5 right out of the gates.
```

Comment:
- Can you comment on the breaking changes of v5 which will be relevant to this project? So we can get a grasp of the additional effort you have to put into this, when compared to developing it for v4.
- Will this solution still be compatible with v4, or will it only work from v5?

Reply:
We skip this in the revised Level 2 setup and will do this in a subsequent effort when v5 is actually released.